synapse/lib/lmdb.py

```diff
@@ -0,0 +1,443 @@
+
+class Seqn:
+    '''
+    An append optimized sequence of byte blobs.
+
+    Args:
+        lenv (lmdb.Environment): The LMDB Environment.
+        name (str): The name of the sequence.
+    '''
+    def __init__(self, lenv, name):
+
+        self.lenv = lenv
+        self.db = lenv.open_db(name)
+
+        with lenv.begin() as xact:
+
+            indx = self.nextindx(xact)
+            self.indx = itertools.count(indx)
+
+    def save(self, xact, items):
+        '''
+        Save a series of items to a sequence.
+
+        Args:
+            xact (lmdb.Transaction): An LMDB write transaction.
+            items (tuple): The series of items to save into the sequence.
+
+        Returns:
+            None
+        '''
+        rows = []
+        for item in items:
+            byts = s_msgpack.en(item)
+            lkey = struct.pack('>Q', next(self.indx))
+            rows.append((lkey, byts))
+
+        with xact.cursor(db=self.db) as curs:
+            curs.putmulti(rows, append=True)
+
+    def nextindx(self, xact):
+        '''
+        Determine the next insert offset according to storage.
+
+        Args:
+            xact (lmdb.Transaction): An LMDB transaction.
+
+        Returns:
+            int: The next insert offset.
+        '''
+        indx = 0
+        with xact.cursor(db=self.db) as curs:
+            if curs.last():
+                indx = struct.unpack('>Q', curs.key())[0] + 1
+        return indx
+
+    def iter(self, xact, offs):
+        '''
+        Iterate over items in a sequence from a given offset.
+
+        Args:
+            xact (lmdb.Transaction): An LMDB transaction.
+            offs (int): The offset to begin iterating from.
+
+        Yields:
+            (indx, valu): The index and valu of the item.
+        '''
+        lkey = struct.pack('>Q', offs)
+        with xact.cursor(db=self.db) as curs:
+
+            if not curs.set_key(lkey):
+                return
+
+            for lkey, lval in curs.iternext():
+                indx = struct.unpack('>Q', lkey)[0]
+                valu = s_msgpack.un(lval)
+                yield indx, valu
+
+class Metrics:
+    '''
+    A helper for recording metrics about an Environment.
+
+    Args:
+        lenv (lmdb.Environment): The LMDB Environment.
+        name (str): The name of the metrics instance.
+    '''
+    def __init__(self, lenv, name=b'metrics'):
+
+        self.lenv = lenv
+
+        self._db_history = lenv.open_db(name + b':history')
+        self._db_current = lenv.open_db(name + b':current')
+
+        self.info = {}
+
+        with lenv.begin() as xact:
+
+            for lkey, lval in xact.cursor(db=self._db_current):
+
+                name = lkey.decode('utf8')
+                valu = struct.unpack('>Q', lval)[0]
+
+                self.info[name] = valu
+
+            indx = 0
+
+            curs = xact.cursor(db=self._db_history)
+            if curs.last():
+                indx = struct.unpack('>Q', curs.key())[0]
```

Comment:
Don't you want to add 1 to the last value? In the above code, the next record will repeat the indx value.

Reply:
note - We need to double check this behavior

Reply:
looking

Reply:
yeah this is busted. @jnwatson good catch

Reply:
fixed